akka-stream-tests/src/test/scala/akka/stream/io/TlsSpec.scala

```diff
@@ -0,0 +1,312 @@
+      def flow(leftClosing: Closing, rightClosing: Closing, flow: Flow[SslTlsInbound, SslTlsOutbound, Any]) = {
+        val local = server(clientTls(rightClosing).reversed join flow)
+        serverTls(leftClosing) join StreamTcp().outgoingConnection(local)
+      }
+    }
+
+    val communicationPatterns =
+      Seq(
+        ClientInitiates,
+        ServerInitiates,
+        ClientInitiatesViaTcp,
+        ServerInitiatesViaTcp)
+
+    trait PayloadScenario extends Named {
+      def flow: Flow[SslTlsInbound, SslTlsOutbound, Any] =
+        Flow[SslTlsInbound]
+          .map {
+            var session: SSLSession = null
+            def setSession(s: SSLSession) = {
+              session = s
+              system.log.debug(s"new session: $session (${session.getId mkString ","})")
+            }
+
+            {
+              case SessionTruncated ⇒ SendBytes(ByteString("TRUNCATED"))
+              case SessionBytes(s, b) if session == null ⇒
+                setSession(s)
+                SendBytes(b)
+              case SessionBytes(s, b) if s != session ⇒
+                setSession(s)
+                SendBytes(ByteString("NEWSESSION") ++ b)
+              case SessionBytes(s, b) ⇒ SendBytes(b)
+            }
+          }
+      def leftClosing: Closing = IgnoreComplete
+      def rightClosing: Closing = IgnoreComplete
+
+      def inputs: immutable.Seq[SslTlsOutbound]
+      def output: ByteString
+
+      protected def send(str: String) = SendBytes(ByteString(str))
+      protected def send(ch: Char) = SendBytes(ByteString(ch.toByte))
+    }
+
+    object SingleBytes extends PayloadScenario {
+      val str = "0123456789"
+      def inputs = str.map(ch ⇒ SendBytes(ByteString(ch.toByte)))
+      def output = ByteString(str)
+    }
+
+    object MediumMessages extends PayloadScenario {
+      val strs = "0123456789" map (d ⇒ d.toString * (rnd.nextInt(9000) + 1000))
+      def inputs = strs map (s ⇒ SendBytes(ByteString(s)))
+      def output = ByteString((strs :\ "")(_ ++ _))
+    }
+
+    object LargeMessages extends PayloadScenario {
+      // TLS max packet size is 16384 bytes
+      val strs = "0123456789" map (d ⇒ d.toString * (rnd.nextInt(9000) + 17000))
+      def inputs = strs map (s ⇒ SendBytes(ByteString(s)))
+      def output = ByteString((strs :\ "")(_ ++ _))
+    }
+
+    object EmptyBytesFirst extends PayloadScenario {
+      def inputs = List(ByteString.empty, ByteString("hello")).map(SendBytes)
+      def output = ByteString("hello")
+    }
+
+    object EmptyBytesInTheMiddle extends PayloadScenario {
+      def inputs = List(ByteString("hello"), ByteString.empty, ByteString(" world")).map(SendBytes)
+      def output = ByteString("hello world")
+    }
+
+    object EmptyBytesLast extends PayloadScenario {
+      def inputs = List(ByteString("hello"), ByteString.empty).map(SendBytes)
+      def output = ByteString("hello")
+    }
+
+    object ImpatientRHS extends PayloadScenario {
```

Comment:
This needs some comments, "impatient" is not descriptive enough. So this takes any bytes it receives, chunks them up in single-byte fragments, takes the first 5 of them, and then passes them on delayed.